### PR TITLE
feat(ark-contracts): create offer

### PR DESCRIPTION
**Pull request by remiroyc**
## Description

Create an offer for a a token 

## What type of PR is this? (check all applicable)

- [X] 🍕 Feature (`feat:`)

## Related Tickets & Documents

https://screenshot.fibery.io/favorites/@my-space/My-Work-435#Product_Management/Issue/Create-offer-127

## Added tests?

- [X] 👍 yes

## Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 Documentation
- [X] 🙅 no documentation needed

### CODE REVIEW

**Review by glihm — CHANGES_REQUESTED** [review]
We've checked with @kwiss and we have to ensure that for a token offer we only have one valid offer for an Account.

For in the storage, @kwiss is adding `token_offers` which is like this:
```rust
// <(token_hash, offerer_address), order_hash>
token_offers: LegacyMap<(felt252, ContractAddress), felt252>)
```

So, when an offer order is created you have to make sure that the offerer does not already have an pending offer for this token hash.

Please check the @kwiss PR #175

**Comment by remiroyc** [comment]
> We've checked with @kwiss and we have to ensure that for a token offer we only have one valid offer for an Account.
> 
> For in the storage, @kwiss is adding `token_offers` which is like this:
> 
> ```rust
> // <(token_hash, offerer_address), order_hash>
> token_offers: LegacyMap<(felt252, ContractAddress), felt252>)
> ```
> 
> So, when an offer order is created you have to make sure that the offerer does not already have an pending offer for this token hash.
> 
> Please check the @kwiss PR #175

It's only applicable to auction-related offers, correct?

File: crates/ark-contracts/arkchain/tests/order/test_order_v1.cairo [hunk 1]
```diff
@@ -11,78 +11,98 @@ use arkchain::order::types::RouteType;
...
 }
 
 #[test]
 fn should_returns_invalid_order_with_zero_token_id() {
     let (order_listing, _, _, _) = setup();
-    let block_timestmap: u64 = 1699556828;
+    let block_timestamp: u64 = 1699556828;
 
     let mut invalid_order = order_listing.clone();
     invalid_order.token_id = 0;
```
Comment (glihm) [inline crates/ark-contracts/arkchain/tests/order/test_order_v1.cairo:47]: `token_id` can be equal to `0`. Also... I am pretty sure that token_id 0 is a valid token_id, no?

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 2]
```diff
@@ -163,10 +168,8 @@ mod orderbook {
                 OrderType::Auction => {
                     self._create_listing_auction(order, order_type, order_hash);
                 },
-                OrderType::Offer => { self._create_listing_offer(order, order_type, order_hash); },
-                OrderType::CollectionOffer => {
-                    self._create_listing_collection_offer(order, order_type, order_hash);
-                },
+                OrderType::Offer => { self.create_offer(order, order_type, order_hash); },
+                OrderType::CollectionOffer => { self.create_offer(order, order_type, order_hash); },
```
Comment (glihm) [inline crates/ark-contracts/arkchain/src/orderbook.cairo:172]: Collection offer and token offer share the same function?

**Review by glihm — COMMENTED** [review]
Some first comments. :+1:

File: crates/ark-contracts/arkchain/src/orderbook.cairo (hunk 2 above)
Comment (remiroyc) [reply crates/ark-contracts/arkchain/src/orderbook.cairo]: Yes, I've just made the token_id optional. Do you think there's something specific to the collection offer that would justify creating a separate function?

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 3]
```diff
@@ -216,22 +226,20 @@ mod orderbook {
...
             ref self: ContractState, order: OrderV1, order_type: OrderType, order_hash: felt252
         ) {
+            // Manage auction offer
+            let ressource_hash = order.compute_ressource_hash();
+            let (auction_order_hash, auction_end_date) = self.auctions.read(ressource_hash);
+            if auction_order_hash.is_non_zero() {
+                let auction_order = self.auction_offers.read(order_hash);
+                assert(auction_order.is_zero(), 'already existing auction offer');
+                self.auction_offers.write(order_hash, auction_order_hash);
+            }
```
Comment (remiroyc) [inline crates/ark-contracts/arkchain/src/orderbook.cairo:240]: @kwiss, could you take a close look at this code to see if I've correctly understood how to use the storage?

File: crates/ark-contracts/arkchain/src/order/order_v1.cairo [hunk 4]
```diff
@@ -103,47 +101,49 @@ impl OrderTraitOrderV1 of OrderTrait<OrderV1> {
...
-        buf.append(felt_token_address);
+    fn compute_token_hash(self: @OrderV1) -> felt252 {
+        let mut buf: Array<felt252> = array![];
+        match self.token_id {
+            Option::Some(token_id) => {
+                buf.append((*token_id.low).into());
+                buf.append((*token_id.high).into());
+            },
+            Option::None => {}
+        }
```
Comment (glihm) [inline crates/ark-contracts/arkchain/src/order/order_v1.cairo:145]:
```suggestion
    fn compute_token_hash(self: @OrderV1) -> felt252 {
        assert(token_id.is_some(), 'Token ID expected for token hash');
        
        let mut buf: Array<felt252> = array![];
        let token_id = token_id.unwrap();
        buf.append((*token_id.low).into());
        buf.append((*token_id.high).into());
        }
```

**Review by glihm — CHANGES_REQUESTED** [review]

**Review by glihm — CHANGES_REQUESTED** [review]

File: crates/ark-contracts/arkchain/src/lib.cairo [hunk 5]
```diff
@@ -10,4 +10,3 @@ mod crypto {
 }
 
 mod orderbook;
-mod tests_lib;
```
Comment (kwiss) [inline crates/ark-contracts/arkchain/src/lib.cairo:13]: why did you remove tests_lib ? it should be included, actually i move the setup function here so we can reuse the setup, can you do that in you test also ? in the same file

Comment (remiroyc) [reply crates/ark-contracts/arkchain/src/lib.cairo:13]: I just moved your setup function into the same directory where the tests are located

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 6]
```diff
@@ -367,10 +372,20 @@ mod orderbook {
                     }
                 );
         }
-        /// Creates a collection offer order.
-        fn _create_collection_offer(
+
+        fn _create_offer(
             ref self: ContractState, order: OrderV1, order_type: OrderType, order_hash: felt252
         ) {
+            // Manage auction offer
```
[the remark this answers is not in the document]
Comment (remiroyc) [reply crates/ark-contracts/arkchain/src/orderbook.cairo:379]: => create_order

File: crates/ark-contracts/arkchain/src/orderbook.cairo [hunk 7]
```diff
@@ -448,10 +456,43 @@ mod orderbook {
...
+                    .auctions
+                    .write(
+                        token_hash,
+                        (
+                            auction_order_hash,
+                            auction_end_date + EXTENSION_TIME_IN_SECONDS,
+                            auction_offer_count + 1
+                        )
+                    );
+            }
```
Comment (remiroyc) [inline crates/ark-contracts/arkchain/src/orderbook.cairo:487]: Add a condition: `auction_end_date - current_block_timestamp < 600` to extend the auction time

**Review by kwiss — DISMISSED** [review]
